vpr/src/place/place_constraints.cpp

```diff
@@ -113,6 +113,25 @@ PartitionRegion constrained_macro_locs(const t_pl_macro& pl_macro) {
     return macro_pr;
 }
 
+void constraints_propagation() {
+    auto& place_ctx = g_vpr_ctx.placement();
+    auto& floorplanning_ctx = g_vpr_ctx.mutable_floorplanning();
+
+    for (auto pl_macro : place_ctx.pl_macros) {
+        if (is_macro_constrained(pl_macro)) {
+            /*
+             * Update the PartitionRegion for the head of the macro
+             * based on the constraints of all blocks contained in the macro
+             */
+            PartitionRegion macro_head_pr = update_macro_pr(pl_macro);
+
+            //Get block id of head macro member and update its PartitionRegion
+            ClusterBlockId blk_id = pl_macro.members[0].blk_index;
+            floorplanning_ctx.cluster_constraints[blk_id] = macro_head_pr;
```

Comment:
Should set the tightest constraint possible on all members of the macro, not just the head.  It gives a stronger invariant to later code (checking any block in a macro is good enough).

Reply:
Should also 
1. set the .locked flag for any blocks that have a floorplan constraint that implies a single location
2. set a floorplan_ctx.cluster_constraints for any block that has its .locked flag set (read in from a cluster-level constraints file / placement file).  That again helps later code -- the cluster_constraints data will encapsulate everything we know about cluster location constraints.
That should probably be done in a separate helper function, which can be called by the top-level propagate_place_constraints() (proposed name) function.

Reply:
That makes sense, I will implement the constraints propagation so that we have the tightest constraints for all members of the macro.

I put the code for setting locked flags on a different pull request - #1759 - because there are other changes I am implementing in that one and it felt like it was getting too big for one pull request.